src/Microsoft.ML.Transforms/Text/WordEmbeddingsTransform.cs

```diff
@@ -324,9 +343,189 @@ public Mapper(WordEmbeddingsTransform parent, Schema inputSchema)
                 _outputType = new VectorType(NumberType.R4, 3 * _parent._currentVocab.Dimension);
             }
 
+            public bool CanSaveOnnx(OnnxContext ctx) => true;
+
             public override Schema.Column[] GetOutputColumns()
                 => _parent.ColumnPairs.Select(x => new Schema.Column(x.output, _outputType, null)).ToArray();
 
+            public void SaveAsOnnx(OnnxContext ctx)
+            {
+                foreach (var (input, output) in _parent.Columns)
+                {
+                    var srcVariableName = ctx.GetVariableName(input);
+                    var schema = _parent.GetOutputSchema(InputSchema);
+                    var dstVariableName = ctx.AddIntermediateVariable(schema[output].Type, output);
+                    SaveAsOnnxCore(ctx, srcVariableName, dstVariableName);
+                }
+            }
+
+            private void SaveAsOnnxCore(OnnxContext ctx, string srcVariableName, string dstVariableName)
+            {
+                // Converts 1 column that is taken as input to the transform into one column of output
+                //
+                // Missing words are mapped to k for finding average, k + 1 for finding min, and k + 2 for finding max
+                // Those spots in the dictionary contain a vector of 0s, max floats, and min floats, respectively
+                //
+                // Symbols:
+                // j: length of latent vector of every word in the pretrained model
+                // n: length of input tensor (number of words)
+                // X: word input, a tensor with n elements.
+                // k: # of words in pretrained model (known when transform is created)
+                // S: word labels, k tensor (known when transform is created)
+                // D: word embeddings, (k + 3)-by-j tensor(known when transform is created). The extra three embeddings
+                //      at the end are used for out of vocab words.
```

Comment:
Please indicate what embedding is used for which out-of-vocab. For example, the (k+1)-th row in D is an all-zero vector, the (k+2)-th row is a vector filled with max float, and so on.

Reply:
You can see this explained in a comment previously on line 366. I believed this is something that is so fundamental to the strategy of the whole model that it made sense to explain it at the very beginning.

---
In reply to: [227993952](https://github.com/dotnet/machinelearning/pull/1249#discussion_r227993952) [](ancestors = 227993952)